### ISSUE TITLE
Error in special cause signal logic--my generalization to arbitrary events has a bug 

### ISSUE THREAD

**Issue by klittle314** [klittle314/WI_Covid_data_Dec_2020#2]
Wondering if there is a reset somehow when there is a sequence of 8 below the midline and the 9th value is below the lower control limit?   I did my best to carefully generalize the find_phase_dates function to handle a generic variable (rather than deaths).  It is possible I made an error in making the edits ;) but I don't think so!

When I use the default extension of 10 days, I get:

![WI Pos Cases two special causes Jan 2021](https://user-images.githubusercontent.com/3459577/105654719-343c6780-5e84-11eb-8940-8cd2db4c0a7f.jpg)

The input series is attached.
[WI_Pos_Cases.xlsx](https://github.com/klittle314/WI_Covid_data_Dec_2020/files/5863403/WI_Pos_Cases.xlsx)

Also, the code says the last phase is Epoch 3 but the limits and center line are exponential!

DIGGING IN:   the fit of the 21 points starting 12/20/20 does NOT lead to a positive significant slope on the log10 scale.   And the additional points are not growing, so slope will only have a higher p value.    It looks like it should be in Epoch 3, but the plot is showing Epoch 2.  The calculation loop says Epoch 3, agreeing with my manual calculation.

The 21 points 12/20/20 to 1/9/21 lead to Phase 3 with no slope.   LCL is 694 and midline is 2163.   THERE ARE NO SIGNALS IN THE DATA 1/10 to 1/24.   But the plotted limits are clearly wrong:   The limits should be LCL = 694 and UCL = 6740.

OK.  It looks like the error is in the EXTENSION logic.   The extension reverts to the exponential model fit and overrides the correct midline and limit calculation.   The code I copied from Mason's function just substitutes the string 'var' for the string 'deaths' in variable names.   I checked just now....(but it is late, want Mason confirmation).   So I think the IHI extension logic has a bug.

```
if (extend_days > 0) {
                # extend midline and ucl/lcl
                data_extend <- data.frame(
                  datex = c(data_var$datex, date_max + seq_len(extend_days))) 
                
                data_extend$serial_day <- as.numeric(difftime(data_extend$datex,
                                                              date_phase_start, 
                                                              units = 'days')) + 1
                #17 January 2021:  how does this handle Epochs 1, 4 and Epoch 3 when we use mean of values?
                midline <- predict(phase_change_result$lm, data_extend)
                
                # need to extend phase_index too
                phase_index <- data_extend$datex >= date_phase_start
              }
            }
            
            phase_days <- as.numeric(difftime(date_phase_end, date_phase_start, units = 'day')) + 1
            
            if (length(midline) > 1) {
              midline <- head(midline[phase_index], phase_days)
            }
            
            ucl     <- midline + (3.14 * mmr)
            lcl     <- midline - (3.14 * mmr)
            
            midline  <- 10 ^ midline
            ucl      <- 10 ^ ucl
            lcl      <- 10 ^ lcl
          }
```

When I set the extension to ZERO, then the correct values are plotted for the last phase.

![POSITIVE_checked_extension_bug](https://user-images.githubusercontent.com/3459577/105661054-1d047680-5e92-11eb-9416-86a89f0782c8.png)

CONCLUSION:  There is a bug in the IHI_Covid_Private function find_phase_dates in the extension logic.  Need to account for exponential growth OR not.  It appears that the extension logic is only correct if the last phase is Epoch 2???

**Comment by modernresearch** [comment]
Hi Kevin, you are right that the epoch 3 case is being treated like epoch 2. I just corrected this in a branch in IHI_Covid_Private: 

https://github.com/klittle314/IHI_Covid_Private/commit/ed92df6c62399b856e82e4ac7b9c9bff7f0d6a00

And a screenshot of the WI series you attached above, run with this correction:

![wisconsin-extension-corrected](https://user-images.githubusercontent.com/16848248/105703866-ddfb1300-5edb-11eb-9b3e-5298c5004bfb.png)

There's no correction required for epochs 1 and 4. The code block you pasted above is within the if/else block for epochs 2 and 3, so it never runs for epochs 1 and 4. That's because for epochs 1 and 4, there's no need to recalculate midline, since it's a constant. By extending date_phase_end, we get an extended midline and limits automatically.

Happy to show you the logic in the code if you want - let me know if a quick meeting would help.

**Comment by klittle314** [comment]
Thx I see the midline distinction for Epochs 2 and 3.   I don't see where the limits are calculated differently, depending on Epoch.  Have to go to a meeting now, back by 10 ET.   May help to have a quick zoom call to confirm.

**Comment by modernresearch** [comment]
The limit calculations are the same for epochs 2 and 3: midline +/- (3.14 * mmr), so I don't think they need to be different. It's just that in epoch 2 midline increases, while in epoch 3 it's a constant or decreasing.

If you meant that you don't see the limits calculated differently depending on whether or not we extend the phase, it still doesn't matter, because the calculation for the limits automatically returns the same length as the midline. So they get calculated after the block for extending the phase.

See lines 438-439: https://github.com/klittle314/IHI_Covid_Private/blob/extend-days-correction/functions.R#L438-L439

Happy to zoom if that doesn't clear it up.

**Comment by klittle314** [comment]
Yes, I figured that out finally.    I see some other things that don't look right, I will need a couple of quiet hours to sort out.   More soon, thanks.
